### PR TITLE
Add support for using ansible-runner

### PR DESCRIPTION
**Pull request by adamruzicka**
This also brings the possibility to run one ansible-runner process per batch which should improve performance.

Requires:
- [x] https://github.com/theforeman/foreman-tasks/pull/386

### CODE REVIEW

**Comment by iNecas** [comment]
As per offline discussion, it would be good to be able to choose which runner implementation to use (classic vs. runner one)

File: app/models/setting/ansible.rb [hunk 1]
```diff
@@ -90,6 +90,16 @@ def load_defaults
               false,
               format(N_('%{cfgmgmt} out of sync disabled'),
                      :cfgmgmt => 'Ansible')
+            ),
+            set(
+              'ansible_implementation',
+              N_('Foreman will run Ansible playbooks using this implementation'),
```
Comment (iNecas) [inline app/models/setting/ansible.rb]: Would be good to mention possible values for this setting in description

Comment (iNecas) [reply app/models/setting/ansible.rb]: Or is the `:collection` supposed to be shown somewhere? I don't see it on my devel setup

**Comment by iNecas** [comment]
@adamruzicka what version of ansible-runner have you tested this with? With 1.0.0, that is in fedora 29, I've seen it running, but it stopped at gathering facts, with 1.2.0, getting from https://releases.ansible.com/ansible-runner/rpm/epel-7-x86_64/, it failed with exit code 250

**Comment by adamruzicka** [comment]
I had ansible-runner installed from pip, no idea what version it was. I can check tomorrow

**Comment by iNecas** [comment]
seems like python2 vs python3 issue, pip3 install ansible-runner seemed to help

**Comment by iNecas** [comment]
Here is the latest issue:

A template like this:

```ansible
---
- hosts: all
  tasks:
    - shell: |
        echo hello
      register: out
    - debug: var=out
```

Ends with output like this:

```
PLAY [all] *********************************************************************
TASK [Gathering Facts] *********************************************************
Exit status: 0
```

When running the runner command directly, I see all tasks running. Digging deeper into what's going on

**Comment by iNecas** [comment]
Funny enough, for some reason, the runner skipped number 5 for me - perhaps not a lucky number for runner :)

```
1-32987fd8-df38-41e0-9b9a-b5aa56121996.json
2-54e1ad30-570d-4ba4-84dc-000000000006.json
3-54e1ad30-570d-4ba4-84dc-00000000000d.json
4-3ce14bbf-3479-4ec7-8ef6-8897237f3d07.json
6-54e1ad30-570d-4ba4-84dc-000000000008.json
7-413c4079-a5bb-455a-8742-b894a41fc377.json
8-54e1ad30-570d-4ba4-84dc-000000000009.json
9-86404b74-fc47-40d3-abf7-e967e7e3a260.json
10-087a6835-9174-458a-b40d-5b73d3bd8142.json
```

**Comment by iNecas** [comment]
Ok, I got to working state with this changes https://github.com/adamruzicka/foreman_ansible/pull/1/files

File: lib/foreman_ansible_core/ansible_runner_task_launcher.rb [hunk 2]
```diff
@@ -0,0 +1,165 @@
+require 'fileutils'
+
+module ForemanAnsibleCore
+  class AnsibleRunnerTaskLauncher < ForemanTasksCore::TaskLauncher::AbstractGroup
+    class AnsibleRunnerRunner < ForemanTasksCore::Runner::Parent
```
Comment (iNecas) [inline lib/foreman_ansible_core/ansible_runner_task_launcher.rb]: I would help readability to have the `AnsibleRunnerRunner` in separate file (the methods between `AnsibleRunnerTaskLauncher` and `AnsibleRunnerRunner` might mix quite easily, as it's not very visible where one class begins and other ends.

Comment (iNecas) [reply lib/foreman_ansible_core/ansible_runner_task_launcher.rb]: Also, perhaps we could have module `ForemanAnsibleCore::Runners`, where there would be both `AnsibleRunner` and `Playbook`, this way, we would get rid of the `AnsibleRunnerRunner`, that looks like a typo

**Comment by iNecas** [comment]
Tested and works well: we need to release new foreman-tasks and bump dependency here before merging

**Comment by iNecas** [comment]
foreman-tasks-core just got released with the required changes: let's bump dependency on foreman-tasks-core 0.3.2 here and rebase to prepare for merge

**Comment by adamruzicka** [comment]
Bumped dependency and rebased

**Comment by iNecas** [comment]
CI is failing due to rex rubocop: we need to get this in first https://github.com/theforeman/foreman_remote_execution/pull/410 to get the CI green

**Comment by adamruzicka** [comment]
[test]

**Comment by adamruzicka** [comment]
[test foreman_ansible]

**Comment by iNecas** [comment]
Thanks @adamruzicka, this has been a ride, but we got there :)

**Comment by ares** [comment]
:clap: thanks guys!

**Comment by adamruzicka** [comment]
:tada: Hell of a ride, thanks for bearing with me

**Pull request merged**